More documentation for objectives API. (#445)
Software-Improvement-Group/sigridci: docs/integrations/sigrid-api-documentation.md

## Available end points

* [Maintainability ratings](#maintainability-ratings)
* [Security and reliability findings](#security-and-reliability-findings)
* [Vulnerable libraries in Open Source Health](#vulnerable-libraries-in-open-source-health)
* [Architecture Quality ratings](#architecture-quality-ratings)
* [Architecture Quality data](#architecture-quality-data)
* [System metadata](#system-metadata)
* [System lifecycle management](#system-lifecycle-management)
* [System and portfolio objectives](#system-objectives)

> 

Reviewer: I'll miss this list (unless there's an auto-generated ToC).

Author: There is! The `<sig-toc>` thingy automatically creates it based on the header tags in the HTML.

</details>

#### Retrieving objectives for a system

    GET https://sigrid-says.com/rest/analysis-results/api/v1/objectives/{customer}/{system}/config
    
This end point returns the following response structure:

    {
      "MAINTAINABILITY": 4.0,
      "NEW_CODE_QUALITY": 3.5,
      "OSH_MAX_SEVERITY": "LOW",
      "TEST_CODE_RATIO": 0.8
    }

Systems always have a single target per objective type. If you define multiple overlapping objectives, 
precedence rules are applied to determine which objective "wins" and gets to decide the target for
that system:

> - System-level objectives have precedence over portfolio-level objectives.
> - Conditional portfolio-level objectives have precedence over unconditional portfolio-level objectives.
> - Unconditional portfolio-level objectives have the lowest precedence, and therefore act like sensible
>   default policies for your portfolio.

Reviewer: This second bullet is what we call "Principle 2". However, it has never been implemented, so I think we should delete this second bullet entirely. Here's my suggestion:
```suggestion
- System-level objectives have precedence over portfolio-level objectives.
- If the metadata of a system matches more than one portfolio objective of the same type, the portfolio objective with the strictest target "wins". 
```

</details>

#### Retrieving objectives for a system

    GET https://sigrid-says.com/rest/analysis-results/api/v1/objectives/{customer}/{system}/config
    
This end point returns the following response structure:

    {
      "MAINTAINABILITY": 4.0,
      "NEW_CODE_QUALITY": 3.5,
      "OSH_MAX_SEVERITY": "LOW",
      "TEST_CODE_RATIO": 0.8
    }

> In Sigrid, a system always has at most one objective per type. So in the above example, the 

Reviewer: I probably suggested this text myself, but I'm now wondering whether it should say: "In Sigrid, a system always has at most one _target_ per type".

Author: I rewrote the entire section a bit. I'm now separating the concept of precedence rules existing from the concrete examples. And yes, I'm intentionally not explaining the situation where you have multiple conditional objectives, I find that makes it a bit too complicated.

This would delete the system-level test code ratio objective, if it exists at the system level, and 
return 204. In case the objective does not exist at the system level, the endpoint returns 404. 

#### Retrieving objectives status

    GET https://sigrid-says.com/rest/analysis-results/api/v1/objectives-evaluation/{customer}?startDate={start}&endDate={end}
    

> This will return the current objectives status and trend for all systems in your portfolio, for the specified time period. The `startDate` and `endDate` parameters accept dates in [ISO 8601 format](https://en.wikipedia.org/wiki/ISO_8601), for example 2024-05-17. 

Reviewer: I prefer "delta" instead of "trend": the former implies exactly two points in time (the start date and the end date), while the latter implies many points in time (a number of time points between the start date and end date):
```suggestion
This will return the current objectives status and delta for all systems in your portfolio, for the specified time period. The `startDate` and `endDate` parameters accept dates in [ISO 8601 format](https://en.wikipedia.org/wiki/ISO_8601), for example 2024-05-17. 
```

This end point returns the following response:

<details markdown="1">
  <summary>Example objectives status response</summary>

```json
{
  "systems": [
    "systemName": "my-example-system",
      "objectives": [
        {
          "type": "OSH_MAX_LICENSE_RISK",
          "feature": "OPEN_SOURCE_HEALTH",
          "target": "LOW",
          "targetMetAtStart": "NOT_MET",
          "targetMetAtEnd": "NOT_MET",
          "delta": "SIMILAR",
          "stateAtEnd": "MEDIUM",
          "level": "PORTFOLIO",
          "parentId": 16
        },
        {
          "type": "ARCHITECTURE_QUALITY",
          "feature": "ARCHITECTURE_QUALITY",
          "target": 4.0,
          "targetMetAtStart": "NOT_MET",
          "targetMetAtEnd": "NOT_MET",
          "delta": "DETERIORATING",
          "stateAtEnd": 3.68909,
          "level": "SYSTEM"
        }
      ]
    ]
  }
}
```

</details>

> The response contains every system in your portfolio, with each system containing all objectives that apply to that system. The objectives follow the same precedence rules as in Sigrid, so only objectives that actually apply will be included in the results.

Reviewer: Suggestion: try to avoid the term "apply". In principle, all portfolio objectives apply to all systems at all time, however, if the metadata of a system doesn't match the condition(s) of the objective, it doesn't result in a target.

This seems like hair splitting, but we already witnessed a case were a user thought that each portfolio objective, next to conditions, had a list of systems to which it applies (that user was asking where they could edit said list).

```suggestion
The response contains every system in your portfolio, with each system containing all objectives that apply to that system. The objectives follow the same precedence rules as in Sigrid, so only objectives with conditions that match the metadata of the system are included in the result.
```
